litmus-portal/cluster-agents/subscriber/pkg/k8s/operations.go

```diff
@@ -2,42 +2,135 @@ package k8s
 
 import (
 	"context"
+	"errors"
 	"fmt"
 	"log"
 	"os"
+	"strings"
+	"sync"
+	"time"
 
 	yaml_converter "github.com/ghodss/yaml"
+	"github.com/sirupsen/logrus"
+	yaml2 "gopkg.in/yaml.v2"
 	corev1 "k8s.io/api/core/v1"
-	"k8s.io/apimachinery/pkg/api/errors"
+	k8s_errors "k8s.io/apimachinery/pkg/api/errors"
 	"k8s.io/apimachinery/pkg/api/meta"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
 	"k8s.io/apimachinery/pkg/runtime/serializer/yaml"
+	memory "k8s.io/client-go/discovery/cached"
 	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/kubernetes"
 	"k8s.io/client-go/restmapper"
-
-	memory "k8s.io/client-go/discovery/cached"
 )
 
 const (
 	ExternAgentConfigName = "agent-config"
+	LiveCheckMaxTries     = 6
 )
 
+type AgentComponents struct {
+	Deployments      []string `yaml:"DEPLOYMENTS"`
+	LiveStatus       bool
+	AccessLiveStatus sync.Mutex
+}
+
 var (
 	Ctx             = context.Background()
 	decUnstructured = yaml.NewDecodingSerializer(unstructured.UnstructuredJSONScheme)
 	dr              dynamic.ResourceInterface
 	AgentNamespace  = os.Getenv("AGENT_NAMESPACE")
 )
 
+func CheckComponentStatus(componentEnv string) error {
+	if componentEnv == "" {
+		return errors.New("components not found in agent config")
+	}
+
+	clientset, err := GetGenericK8sClient()
+	if err != nil {
+		return err
+	}
+
+	var components AgentComponents
+
+	err = yaml2.Unmarshal([]byte(strings.TrimSpace(componentEnv)), &components)
+	if err != nil {
+		return err
+	}
+
+	components.LiveStatus = true
+	components.AccessLiveStatus = sync.Mutex{}
+
+	wait := sync.WaitGroup{}
+
+	// add all agent components to waitgroup
+	wait.Add(1)
+	go checkDeploymentStatus(&components, clientset, &wait)
+
+	wait.Wait()
+	if !components.LiveStatus {
+		return errors.New("all components failed to startup")
+	}
+	return nil
+}
+
+func checkDeploymentStatus(components *AgentComponents, clientset *kubernetes.Clientset, wait *sync.WaitGroup) {
+	downCount := 0
+	retries := 0
+	defer wait.Done()
+	for retries < LiveCheckMaxTries {
+		for _, dep := range components.Deployments {
+			podList, err := clientset.CoreV1().Pods(AgentNamespace).List(metav1.ListOptions{LabelSelector: dep})
+			if err != nil {
+				logrus.Errorf("failed to get deployment pods %v , err : %v", dep, err.Error())
+				downCount += 1
+				continue
+			}
+			if len(podList.Items) == 0 {
+				logrus.Errorf("failed to get deployments pods %v , err : pods not found", dep)
+				downCount += 1
+				continue
+			}
+		outer:
+			for _, pod := range podList.Items {
+				if pod.Status.Phase != corev1.PodRunning {
+					logrus.Errorf("failed to get running pods for dep %v", dep)
+					downCount += 1
+					break
```

Comment:
do you mean continue?

Reply:
No break because i any of the pods/containers are not running I break out from the list as we need all the pods to be up